Feature/issue 3303 distinct aggr support changelog

yokofly opened the pull request:
Please write user-readable short description of the changes:
- make distinct this aggr combinator support changelog/retract.
- the basic design uses a hashmap(internal is using btreemap) to save the state(count value).  and pass the processed data to the next function.
- also enhance changelog function parse.
- add some helper function in CountedValueMap(internal btreemap)

test:
count_distinct_if shall be found in https://github.com/timeplus-io/proton/compare/feature/issue-3303-distinct-aggr-support-changelog?expand=1#diff-f3bf2509e030a3c87801df81d8e1b41cc2062a0bd3930eca45a8026cbb88320aR214-R228
serialize ckpt also works, see https://github.com/timeplus-io/proton/compare/feature/issue-3303-distinct-aggr-support-changelog?expand=1#diff-f3bf2509e030a3c87801df81d8e1b41cc2062a0bd3930eca45a8026cbb88320aR36-R43

and we have reported 2 issues, one is still open #425 , one(#419) will be fixed in commit 2fea686cba3de5bb6af5caaf3b9b3a040a3e0a81

yokofly committed:
fix typo aggregation instead `aggreagtion`

part 4(there still leave a bug in count_distinct, but sum_distinct work well). also the parse change meaningless.
todo: fix bug && revert parse change && add normal test and complex count(distinct (if ...) such style test.

yokofly committed:
fix https://github.com/timeplus-io/proton/issues/419
we shall do (de)serialize and merge for nested func as well.

yokofly committed:
finish dev:
1. add helper Fn in CountedValueMap
2. change the parse position from ExpressionAnalyzer to
substituteStreamingFunc
3. current retract do not optimize for the single numeric input
4. add some necessary testing

yokofly committed:
Merge remote-tracking branch 'origin/develop' into feature/issue-3303-distinct-aggr-support-changelog

yokofly committed:
fix typo and enforce at most 2 arguments in count_retract

yl-lisen commented on src/Interpreters/Streaming/SubstituteStreamingFunction.cpp:
```diff
@@ -169,6 +173,13 @@ void StreamingFunctionData::visit(DB::ASTFunction & func, DB::ASTPtr)
+        else
+        {
+            /// replace `<aggr>_distinct[_combinator]` ==> `<aggr>_distinct_streaming[_combinator]` for streaming query
+            constexpr std::string_view distinct_raw{"_distinct"}, distinct_streaming_raw{"_distinct_streaming"};
+            if (size_t pos = func.name.find(distinct_raw); pos != std::string::npos)
+                func.name.replace(pos, distinct_raw.length(), distinct_streaming_raw);
+        }
```
/// Always show original func name
func.covered_name = func.name;

yl-lisen replied:
we should define a function:
```c++
void ASTFunction::substitute(String new_name)
{
    /// Always show original func name
    if (func.covered_name.empty())
        func.covered_name = func.name;

    func.name = std::move(new_name);
}
```

yl-lisen replied:
and please use this func instead of anywhere needs substitute func ,
some one is in `StreamingNowFunctionData::visit`

yl-lisen commented on tests/stream/test_stream_smoke/0013_changelog_stream13.yaml:
```diff
@@ -0,0 +1,263 @@
+        query_type: stream
+        depends_on_stream: v_12181627
+        query_id: '12181627213'
+        query: subscribe to select sum_distinct(val), sum(val), count_distinct(val), count(val)
+          from v_12181627 settings checkpoint_interval=1;
```
Another scenario is global aggregation emit changelog, for example:
```sql
subscribe to select count_distinct(val) from v emit changelog setting checkpoint_interval = 1
```

yl-lisen commented on src/Interpreters/Streaming/SubstituteStreamingFunction.cpp:
```diff
@@ -77,21 +76,26 @@ std::optional<String> StreamingFunctionData::supportChangelog(const String & fun
+    /// `count_distinct`        => `__count_retract_distinct_retract`
+    /// `count_distinct_if`     => `__count_retract_distinct_retract_if`
     String combinator_suffix;
+    constexpr std::string_view distinct_raw{"_distinct"}, distinct_retract_raw{"_distinct_retract"};
```
directly use literal string "_distinct" is better ?

yl-lisen commented on src/Interpreters/Streaming/SubstituteStreamingFunction.cpp:
```diff
@@ -159,7 +163,7 @@ void StreamingFunctionData::visit(DB::ASTFunction & func, DB::ASTPtr)
                     /// SELECT count(), avg(i) FROM (SELECT i, _tp_delta FROM versioned_kv) GROUP BY i; =>
                     /// SELECT __count_retract(_tp_delta), __avg_retract(i, _tp_delta) FROM (SELECT i, _tp_delta FROM versioned_kv) GROUP BY i; =>
-                    if (func.name.starts_with("__count_retract") && delta_pos - func.arguments->children.begin() > 0)
+                    if ((func.name == "__count_retract" || func.name == "__count_retract_if") && delta_pos - func.arguments->children.begin() > 0)
                         /// Fix for nullable since this substitution is not equal
                         func.arguments->children[0] = std::make_shared<ASTIdentifier>(ProtonConsts::RESERVED_DELTA_FLAG);
                     else
```
we can remove this special path

yl-lisen commented on src/AggregateFunctions/Streaming/CountedValueMap.h:
```diff
@@ -38,6 +38,7 @@ struct CountedValueArena<StringRef>
             arena.free(const_cast<char *>(key.data), key.size);
     }
 
+    ArenaWithFreeLists * getArenaWithFreeLists() { return &arena; }
 private:
```
separate with an empty line

yl-lisen commented on src/AggregateFunctions/Streaming/CountedValueMap.h:
```diff
@@ -98,6 +99,35 @@ class CountedValueMap
         return true;
     }
 
+    /// return true meaning: this is a new added element
+    bool insertIfNewData(T v)
+    {
+        if (atCapacity()) [[unlikely]]
+        {
+            /// At capacity, this is an optimization
+            /// fast ignore elements we don't want to maintain
+            if (less(lastValue(), v))
+                return false;
+        }
+
+        auto iter = m.find(v);
+        if (iter != m.end())
+        {
+            ++iter->second;
+        }
+        else
+        {
+            /// Didn't find v in the map
+            [[maybe_unused]] auto [_, inserted] = m.emplace(arena->emplace(std::move(v)), 1);
+            assert(inserted);
+
+            eraseExtraElements();
+            return true;
+        }
```
we shouldn't copy another same code, we can refactor this interface:
```c++
/// \returns: the emplaced element iterator, if failed to emplace, return invalid iterator, `m.end()`
iter tryEmplace(T v)
{
    ...
    return <iter> or <empty_iter>;
}

/// \returns: added.
bool insert(T v)
{
    auto iter = tryEmplace(std::move(v));
    return iter != m.end();
}
```

yl-lisen replied:
For your distinct implementation, you can use
```
auto iter = map.tryEmplace(key);
/// If added and ref count is 1, assume it's a new key
if (iter != map.end() && iter->second == 1)
{
    ....
}
```

yl-lisen replied:
same for erase

yl-lisen commented on src/AggregateFunctions/Streaming/AggregateFunctionDistinctRetract.h:
```diff
@@ -0,0 +1,341 @@
+                for (size_t i = 0; i < argument_size - 1; ++i)
+                    begin = argument_columns[i]->deserializeAndInsertFromArena(begin);
+
+                argument_columns[argument_size - 1]->insert(static_cast<int8_t>(+1));
```
argument_columns.back()->insert(Int8(1));

yl-lisen commented on src/AggregateFunctions/Streaming/AggregateFunctionDistinctRetract.h:
```diff
@@ -0,0 +1,341 @@
+                    begin = argument_columns[i]->deserializeAndInsertFromArena(begin);
+
+                /// insert delta_col
+                argument_columns[argument_size - 1]->insert(delta_col);
```
argument_columns.back()->insert(delta_col);

yokofly committed:
fix review comment:
0. extra a new ASTFunction::substitute(new name) function (proton only)
1. polish dup code in CountedValueMap
2. modify the test time to pass CI

yokofly committed:
fix format

yokofly commented on src/Interpreters/Streaming/SubstituteStreamingFunction.cpp:
```diff
@@ -177,8 +176,13 @@ void StreamingFunctionData::visit(DB::ASTFunction & func, DB::ASTPtr)
+            if (size_t pos = func.name.find(distinct_raw); pos != std::string::npos) {
+                std::string new_name = func.name;
+                new_name.replace(pos, distinct_raw.length(), distinct_streaming_raw);
+
```
this is a bit ugly, but I do not wanna introduce boost..
```cpp
if (string.contains("_distinct")) // cpp 23 , proton use cpp@20
if (str.find("_distinct") != std::string::npos) // current proton style
{
    auto new_name(func.name);
    boost::replace_first(new_name, "_distinct", "_distinct_streaming")
    func.substitute(new_name)
 }
 ```

yokofly replied:
wait, proton has using cpp 23, but the `replace` still manual in cpp standard
edit: ummm but proton uses cpp20

yl-lisen commented on src/AggregateFunctions/Streaming/CountedValueMap.h:
```diff
@@ -72,48 +73,28 @@ class CountedValueMap
     }
 
-    bool insert(T v)
+    /**
+    - If returnIfNewData is true, 
+        the function behaves returning true only if the element is new insert instead.
+    - If returnIfNewData is false (which is the default), 
+        the function behaves like the original insert,
+        returning true in all cases except when the capacity is full, and the new element is not desired.
+    */
```
For a good interface, it should have a clear parameters and returns.
but now, it's complex and confusing

yokofly replied:
is below works?
```cpp

    /**
    * Inserts an element into the map.
    * 
    * @param v The element to be inserted.
    * @param returnTrueIfNewData Controls the return behavior:
    *    - If true, returns true only if a new element is actually inserted.
    *    - If false (default), behaves like the standard insert: 
    *      returns true except when the capacity is full and the new element is not inserted.
    * 
    * @return True if the element is successfully processed (inserted or found), false otherwise.
    */
    bool insert(T v, bool returnTrueIfNewData = false)
    {
        if (atCapacity())
        {
            /// At capacity, this is an optimization
            /// fast ignore elements we don't want to maintain
            if (less(lastValue(), v))
                return false;
        }

        auto iter = m.find(v);
        if (iter != m.end())
        {
            ++iter->second;
            return !returnTrueIfNewData;
        }
        else
        {
            /// Didn't find v in the map
            [[maybe_unused]] auto [_, inserted] = m.emplace(arena->emplace(std::move(v)), 1);
            assert(inserted);

            eraseExtraElements();
            return true;
        }
    }

    /**
    * Erases an element from the map, if present.
    * 
    * @tparam TT The type of the element to be erased.
    * @param v The element to be erased.
    * @param returnTrueIfNewErased Controls the return behavior:
    *    - If true, returns true only if the element is found and actually erased (i.e., its count reaches zero).
    *    - If false (default), returns true if the element is found, regardless of whether it's actually erased.
    * 
    * @return True if the element is successfully processed based on the 'returnTrueIfNewErased' flag, false otherwise.
    */
    template <typename TT>
    bool erase(const TT & v, bool returnTrueIfNewErased = false) /// To enable heterogeneous erase
    {
        auto iter = m.find(v);
        if (iter != m.end())
        {
            --iter->second;
            if (iter->second == 0)
            {
                arena->free(iter->first);
                m.erase(iter);
                return true;
            }
            return !returnTrueIfNewErased;
        }
        return false;
    }

```

yokofly committed:
Merge remote-tracking branch 'origin/develop' into feature/issue-3303-distinct-aggr-support-changelog

yokofly committed:
Merge remote-tracking branch 'origin/develop' into feature/issue-3303-distinct-aggr-support-changelog

yokofly committed:
fix review comment:
1. change the counted value map helper function(name conversion and design guard)
2. add extra test with obvious naming like(sum_distinct_streaming)
3. clean trash in parse

yl-lisen commented on src/Interpreters/Streaming/SubstituteStreamingFunction.cpp:
```diff
@@ -153,19 +158,25 @@ void StreamingFunctionData::visit(DB::ASTFunction & func, DB::ASTPtr)
+        else
+        {
+            /// replace `<aggr>_distinct[_combinator]` ==> `<aggr>_distinct_streaming[_combinator]` for streaming query
+            if (!boost::algorithm::contains(func.name, "_distinct_streaming")) {
+                if (boost::algorithm::contains(func.name, "_distinct")) {
+                    std::string new_name = func.name;
```
```c++
if (boost::algorithm::contains(func_name_lower, "_distinct"))
{
    boost::algorithm::replace_first(func_name_lower, "_distinct", "_distinct_streaming")
    substitueFunction(func, std::move(func_name_lower));
}
```

yokofly replied:
`distinct_if` ?

yokofly replied:
user input: distinct_streaming_if 
we shall skip the change?
so i have to check not contain streaming ?

yokofly replied:
https://godbolt.org/z/xfh1PhKPx  no need to consider user input our internal function naming?
like `distinct_streaming_if`

yl-lisen commented on src/Interpreters/Streaming/SubstituteStreamingFunction.cpp:
```diff
@@ -77,21 +78,25 @@ std::optional<String> StreamingFunctionData::supportChangelog(const String & fun
+            std::string combinator_name = combinator->getName();
             /// TODO: support more combinators
-            if (combinator_name != "_if")
+            if (combinator_name != "_if" && combinator_name != "_distinct" && combinator_name != "_distinct_retract")
```
not need care `_distinct_retract` ， can't happen here

yokofly replied:
what if user input `_distinct_retract` ? do not consider this condition?

yl-lisen replied:
we should block them

yl-lisen replied:
yes, let's keep your changes

yokofly committed:
add `likely` and change the if condition position
- first check contains("distinct") then check not contains("distinctstreaming")

yokofly committed:
Merge remote-tracking branch 'origin/develop' into feature/issue-3303-distinct-aggr-support-changelog

yokofly committed:
user func_name_lower and add a new check with retract in appendonly mode

yokofly committed:
Merge branch 'develop' into feature/issue-3303-distinct-aggr-support-changelog

yokofly committed:
revert a change to pass CI(still specific check the count_retract(*) )

yokofly committed:
add delay time to pass CI

yokofly committed:
add delay time to pass CI
